kernel.rb

```diff
@@ -150,6 +150,44 @@ def yield_self
 
   module_function
 
+  # call-seq:
+  #    loop { block }
+  #    loop            -> an_enumerator
+  #
+  # Repeatedly executes the block.
+  #
+  # If no block is given, an enumerator is returned instead.
+  #
+  #    loop do
+  #      print "Input: "
+  #      line = gets
+  #      break if !line or line =~ /^q/i
+  #      # ...
+  #    end
+  #
+  # StopIteration raised in the block breaks the loop.  In this case,
+  # loop returns the "result" value stored in the exception.
+  #
+  #    enum = Enumerator.new { |y|
+  #      y << "one"
+  #      y << "two"
+  #      :ok
+  #    }
+  #
+  #    result = loop {
+  #      puts enum.next
+  #    } #=> :ok
+  def loop
+    unless Primitive.block_given_p
+      return Primitive.cexpr! 'SIZED_ENUMERATOR(self, 0, 0, rb_f_loop_size)'
```

Comment:
What if we use `Enumerator.produce { nil }` here? Would that be slower than calling a primitive?

Reply:
We don't have a benchmark for this path yet, so I just chose the same way as `#then`, which seems least likely to slow down. We'd need a benchmark if we want to change the way things work here. Do you have an idea for benchmarking this in a realistic manner? Honestly I have never used `#loop` in a way that this path is used.

Reply:
Also, how is it supposed to be used? https://github.com/ruby/ruby/pull/6983/commits/5b56ab1b990bba7ac02f187feb2dbb9f2edb4198 didn't pass CI.

Reply:
Weird, I can check the failure. Normally `Enumerator.produce { nil }` should return an endless stream of `nil` values, which is what a bare `loop` does, as well.

Reply:
Ok, so apparently the enumerator returned by `loop` is supposed to yield nothing, not even `nil`, which is what the spec is testing for. On the other hand `Enumerator.produce { }` or `Enumerator.produce { nil }` always yields `nil` values.

What works though is:
```ruby
return enum_for(:loop) { Float::INFINITY }
```

Reply:
That worked. thank you!

Reply:
FYI TruffleRuby & Rubinius ave this figured out for a long time: https://github.com/oracle/truffleruby/blob/master/src/main/ruby/truffleruby/core/kernel.rb#L407-L418
Feel free to copy Ruby code from there.

Reply:
Nice, I didn't notice the rescue region can be limited to the latter portion.